src/main/java/de/tum/cit/aet/artemis/core/service/LLMTokenUsageService.java

```diff
@@ -0,0 +1,133 @@
+     *
+     * @param llmRequests     List of LLM requests containing details about the token usage.
+     * @param serviceType     Type of the LLM service (e.g., IRIS, GPT-3).
+     * @param builderFunction A function that takes an LLMTokenUsageBuilder and returns a modified LLMTokenUsageBuilder.
+     *                            This function is used to set additional properties on the LLMTokenUsageTrace object, such as
+     *                            the course ID, user ID, exercise ID, and Iris message ID.
+     *                            Example usage:
+     *                            builder -> builder.withCourse(courseId).withUser(userId)
+     * @return The saved LLMTokenUsageTrace object, which includes the details of the token usage.
+     */
+    // TODO: this should ideally be done Async
+    public LLMTokenUsageTrace saveLLMTokenUsage(List<LLMRequest> llmRequests, LLMServiceType serviceType, Function<LLMTokenUsageBuilder, LLMTokenUsageBuilder> builderFunction) {
+        LLMTokenUsageTrace llmTokenUsageTrace = new LLMTokenUsageTrace();
+        llmTokenUsageTrace.setServiceType(serviceType);
+
+        LLMTokenUsageBuilder builder = builderFunction.apply(new LLMTokenUsageBuilder());
+        builder.getIrisMessageID().ifPresent(llmTokenUsageTrace::setIrisMessageId);
+        builder.getCourseID().ifPresent(llmTokenUsageTrace::setCourseId);
+        builder.getExerciseID().ifPresent(llmTokenUsageTrace::setExerciseId);
+        builder.getUserID().ifPresent(llmTokenUsageTrace::setUserId);
+
+        llmTokenUsageTrace.setLlmRequests(llmRequests.stream().map(LLMTokenUsageService::convertLLMRequestToLLMTokenUsageRequest)
+                .peek(llmTokenUsageRequest -> llmTokenUsageRequest.setTrace(llmTokenUsageTrace)).collect(Collectors.toSet()));
+
+        return llmTokenUsageTraceRepository.save(llmTokenUsageTrace);
+    }
+
+    private static LLMTokenUsageRequest convertLLMRequestToLLMTokenUsageRequest(LLMRequest llmRequest) {
+        LLMTokenUsageRequest llmTokenUsageRequest = new LLMTokenUsageRequest();
+        llmTokenUsageRequest.setModel(llmRequest.model());
+        llmTokenUsageRequest.setNumInputTokens(llmRequest.numInputTokens());
+        llmTokenUsageRequest.setNumOutputTokens(llmRequest.numOutputTokens());
+        llmTokenUsageRequest.setCostPerMillionInputTokens(llmRequest.costPerMillionInputToken());
+        llmTokenUsageRequest.setCostPerMillionOutputTokens(llmRequest.costPerMillionOutputToken());
+        llmTokenUsageRequest.setServicePipelineId(llmRequest.pipelineId());
+        return llmTokenUsageRequest;
+    }
+
+    // TODO: this should ideally be done Async
+    public void appendRequestsToTrace(List<LLMRequest> requests, LLMTokenUsageTrace trace) {
+        var requestSet = requests.stream().map(LLMTokenUsageService::convertLLMRequestToLLMTokenUsageRequest).peek(llmTokenUsageRequest -> llmTokenUsageRequest.setTrace(trace))
+                .collect(Collectors.toSet());
```

Comment:
Is there a reason that we collect the requests to a `Set` but not to a `List`?

Reply:
I think you can use either